tinytuya/Cloud.py

```diff
@@ -386,3 +451,46 @@ def getconnectstatus(self, deviceid=None):
                     "Error from Tuya Cloud: %r" % response_dict['msg'],
             )
         return(response_dict["result"]["online"])
+
+    def getdevicelog(self, devid, start=None, end=None, evtype=None, size=100, params={}):

```

Comment:
This default value for `params` is a mutable static variable (refer <https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments>), and as a result carries over its values from the previous request within the same python program, resulting in responses with zero entries when requesting a second device's log after previously retrieving the logs for another device.

Reply:
Thanks @klightspeed !  Pylint also caught this and I forgot to ask @uzlonewolf .  

@uzlonewolf what if set this default to None and add a...
```python
if not params:
     params = {}
```

Reply:
It looks like the
```py
        if type(params) != dict:
            params = {}
```
should already cover if the default value for `params` were `None`.
```py
>>> type(None) != dict
True
```

Reply:
Which leads to the question, since the dict is modified, should we be doing `params = params.copy()` as well?  Keeping the start/end times the same between requests could be useful (or confusing), but changing the size or start_row_key is not.  We could also save/restore just those keys.

I'm leaning towards leaving the default as it is and instead going
```python
if type(params) != dict:
    params = {}
else:
    params = params.copy()
```

Reply:
This hurt my head a bit. I had to run a few test to prove this out and educate myself. :)

In the current state, when getdevicelog() is called.. 

* without params provide, it will set params with the keys (e.g. start_time) that will leak into the global/calling scope so that when it is called again w/o params, those previous values will still be there (the static aspect you talk about).  I'm not sure why we would want this (agree with "confusing").  Can you think of a reason why we would want that?  I can't but I may be missing something.  In any case, both recommendations above will remove this scope leaking. 
* with params specified, it will only modify the params sent and they will not leak to others.  I did some testing, and what I see is that if you provide explicit (static) params (eg. {'id': 123} ), the additional keys added are not leaked back to the calling scope.  However, if you call with a variable that contains the params dict, the additional updates are preserved (in the variable) and will be seen in the calling scope.

So both the recommendation from @klightspeed and @uzlonewolf result in preserving the second point above and fixing the first point.  If no params are specified in the calling scope, nothing is leaked back.

@uzlonewolf I know you are leaning toward leaving the `params={}` default, but if we just change it to `params=None`, we don't even need to add your conditional setting.  The existing conditionals will handle setting params to a empty dict if no params were specified.

My testing:

```python
def f(params={}):
    print("called with %r" % params)
    params['id'] = 1234

def g(params=None):
    if type(params) != dict:
        params = {}
    print("called with %r" % params)
    params['id'] = 1234
```

```python
# EXISTING

# try with default - see the leak
>>> f()
called with {}
>>> f()
called with {'id': 1234}

# try with static params - no leak until empty params is called
>>> f({'z': 99})
called with {'z': 99}
>>> f({'z': 99})
called with {'z': 99}
>>> f()
called with {'id': 1234}

# params as a variable - expected leak until empty params is called
>>> p = {'z': 99}
>>> f(p)
called with {'z': 99}
>>> f(p)
called with {'z': 99, 'id': 1234}
>>> f()
called with {'id': 1234}

# PROPOSED

# default - no leak
>>> g()
called with {}
>>> g()
called with {}

# with static - no leak
>>> g({'z': 99})
called with {'z': 99}
>>> g({'z': 99})
called with {'z': 99}
>>> g()
called with {}

# params as a variable - expected leak but no leak when empty params is called
>>> p = {'z': 99}
>>> g(p)
called with {'z': 99}
>>> g(p)
called with {'z': 99, 'id': 1234}
>>> g()
called with {}
```

Reply:
> we don't even need to add your conditional setting

We do if we don't want `start_row_key` to leak back in the case of `params` being provided and multiple calls are needed to get all requested records, or `size` to get added unexpectedly.

```python
p = {}
c.getdevicelog( ..., start_row_key="...", parms=p) # if either start_row_key is provided or this needs multiple requests to fetch all rows, then
print(p) # this now contains p['start_row_key'] and
c.getdevicelog( ..., parms=p) # will unexpectedly try and start with the above start_row_key
```
Same deal with `size`
```python
p = {}
c.getdevicelog( ..., size=1, parms=p)
print(p) # this now contains p['size'] = 1 and
c.getdevicelog( ..., size=100, parms=p) # will unexpectedly retrieve at most 1 record per call, resulting in 50 cloud calls and (at most) 50 records returned
```

Reply:
Ha!  You're right!  I thought about it just as I was going to sleep last night that we actually may not want this to leak back into the passed variable. You just proved it.

I'll submit a change.  Thanks for spotting this @klightspeed and thanks for finding/fixing he real issue @uzlonewolf .  Brilliant work you two.